sdk/keyvault/internal/challenge_policy.go

```diff
@@ -7,139 +7,101 @@
-		mainResource:            temporal.NewResource(acquire),
+	kv := keyVaultAuthorizer{
 		verifyChallengeResource: !opts.DisableChallengeResourceVerification,
 	}
+	return runtime.NewBearerTokenPolicy(cred, nil, &policy.BearerTokenOptions{
+		AuthorizationHandler: policy.AuthorizationHandler{
+			OnRequest:   kv.authorize,
+			OnChallenge: kv.authorizeOnChallenge,
+		},
+	})
 }
 
-func (k *KeyVaultChallengePolicy) Do(req *policy.Request) (*http.Response, error) {
-	as := acquiringResourceState{
-		p:   k,
-		req: req,
-	}
-
-	if k.scope == nil || k.tenantID == nil {
-		// First request, get both to get the token
-		challengeReq, err := k.getChallengeRequest(*req)
-		if err != nil {
-			return nil, err
-		}
-
-		resp, err := challengeReq.Next()
-		if err != nil {
-			return nil, err
-		}
-
-		if resp.StatusCode > 399 && resp.StatusCode != http.StatusUnauthorized {
-			// the request failed for some other reason, don't try any further
-			return resp, nil
+func (k *keyVaultAuthorizer) authorize(req *policy.Request, authNZ func(policy.TokenRequestOptions) error) error {
+	if len(k.tro.Scopes) == 0 || k.tenantID == "" {
+		if body := req.Body(); body != nil {
+			// We don't know the scope or tenant ID because we haven't seen a challenge yet. We elicit one now by sending
+			// the request without authorization, first removing its body, if any. authorizeOnChallenge will reattach the
+			// body, authorize the request, and send it again.
+			rb := reqBody{body, req.Raw().Header.Get("content-type")}
+			req.SetOperationValue(rb)
+			if err := req.SetBody(nil, ""); err != nil {
+				return err
+			}
 		}
-		err = k.findScopeAndTenant(resp, req.Raw())
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	tk, err := k.mainResource.Get(as)
-	if err != nil {
-		return nil, err
+		// returning nil indicates the bearer token policy should send the request
+		return nil
 	}
+	// else we know the auth parameters and can authorize the request as normal
+	return authNZ(k.tro)
+}
 
-	req.Raw().Header.Set(
-		headerAuthorization,
-		fmt.Sprintf("%s%s", bearerHeader, tk.Token),
-	)
-
-	// send a copy of the request
-	cloneReq := req.Clone(req.Raw().Context())
-	resp, cloneReqErr := cloneReq.Next()
-	if cloneReqErr != nil {
-		return nil, cloneReqErr
+func (k *keyVaultAuthorizer) authorizeOnChallenge(req *policy.Request, res *http.Response, authNZ func(policy.TokenRequestOptions) error) error {
+	// parse the challenge
+	if err := k.updateTokenRequestOptions(res, req.Raw()); err != nil {
+		return err
 	}
-
-	// If it fails and has a 401, try it with a new token
-	if resp.StatusCode == 401 {
-		// Force a new token
-		k.mainResource.Expire()
-
-		// Find the scope and tenant again in case they have changed
-		err := k.findScopeAndTenant(resp, req.Raw())
-		if err != nil {
-			// Error parsing challenge, doomed to fail. Return
-			return resp, cloneReqErr
+	// reattach the request's body, if it was removed by authorize()
+	var rb reqBody
+	if req.OperationValue(&rb) {
```

Comment:
If this doesn't find the thing, what will happen?

Reply:
Then we send the request without a body. If it's e.g. a GET, no problem; if it's a POST, we get an error response from Key Vault. That should be a 400, so non-retriable and therefore an application error. Calling the client method again should lead to success because the policy won't remove the body again. Do you think we should check the verb to decide whether the request should have had a body so we can return a better error pointing to an azcore bug?

Reply:
Hmm we could, but I suspect this code won't change much so maybe it's more effort than it's worth (I'll let you decide on that part).  It might be good to add a comment about this though for future maintainers.

Reply:
I added a comment because I don't want this policy to have an opinion about which requests should have a body.